### PR TITLE
User IUser::getEMailAddress() all over the place

### PR DESCRIPTION
**Pull request by DeepDiver1975**
Please review @nickvergessen @LukasReschke @MorrisJobke @blizzz @schiesbn THX

### CODE REVIEW

**Comment by LukasReschke** [comment]
> 06:06:41 ....................................PHP Fatal error:  Uncaught Error: Call to a member function getEMailAddress() on string in /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/lib/private/share/mailnotifications.php:83

:bomb:

File: core/ajax/share.php [hunk 1]
```diff
@@ -135,17 +135,20 @@
...
-				$recipientList[] = $recipient;
+				$recipientList[] = $userManager->get($recipient);
 			} elseif ($shareType === \OCP\Share::SHARE_TYPE_GROUP) {
 				$recipientList = \OC_Group::usersInGroup($recipient);
+				$group = \OC::$server->getGroupManager()->get($recipient);
+				$recipientList = $group->searchUsers('');
 			}
 			// don't send a mail to the user who shared the file
-			$recipientList = array_diff($recipientList, array(\OCP\User::getUser()));
+			$recipientList = array_diff($recipientList, [$userManager->get(\OCP\User::getUser())]);
```
Comment (MorrisJobke) [inline core/ajax/share.php]:
```json
{"reqId":"ER23TF3i0WggDbkBipdt","remoteAddr":"::1","app":"PHP","message":"
Object of class OC\\User\\User could not be converted to string at master\/core\/ajax\/share.php#148","level":0,"time":"2015-12-02T08:55:13+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
```

When share a file with a owncloud user and try to send a notification email.

Comment (DeepDiver1975) [inline core/ajax/share.php]: > When share a file with a owncloud user and try to send a notification email.

@MorrisJobke fixed - THX for testing

**Comment by MorrisJobke** [comment]
@DeepDiver1975 I still get following:

```
{"reqId":"PlvOvONgwS52eFFC3gjn","remoteAddr":"::1","app":"PHP","message":"Object of class OC\\User\\User could not be converted to string at \/Users\/morrisjobke\/Projects\/owncloud\/master\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Driver\/PDOStatement.php#91","level":0,"time":"2015-12-02T11:17:54+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
{"reqId":"PlvOvONgwS52eFFC3gjn","remoteAddr":"::1","app":"PHP","message":"Object of class OC\\User\\User to string conversion at \/Users\/morrisjobke\/Projects\/owncloud\/master\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Driver\/PDOStatement.php#91","level":0,"time":"2015-12-02T11:17:54+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
{"reqId":"PlvOvONgwS52eFFC3gjn","remoteAddr":"::1","app":"PHP","message":"Illegal offset type in isset or empty at \/Users\/morrisjobke\/Projects\/owncloud\/master\/lib\/private\/user\/manager.php#129","level":0,"time":"2015-12-02T11:17:54+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
{"reqId":"PlvOvONgwS52eFFC3gjn","remoteAddr":"::1","app":"PHP","message":"Illegal offset type in isset or empty at \/Users\/morrisjobke\/Projects\/owncloud\/master\/lib\/private\/user\/database.php#209","level":0,"time":"2015-12-02T11:17:54+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
{"reqId":"PlvOvONgwS52eFFC3gjn","remoteAddr":"::1","app":"PHP","message":"Object of class OC\\User\\User could not be converted to string at \/Users\/morrisjobke\/Projects\/owncloud\/master\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Driver\/PDOStatement.php#91","level":0,"time":"2015-12-02T11:17:54+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
{"reqId":"PlvOvONgwS52eFFC3gjn","remoteAddr":"::1","app":"PHP","message":"Object of class OC\\User\\User to string conversion at \/Users\/morrisjobke\/Projects\/owncloud\/master\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Driver\/PDOStatement.php#91","level":0,"time":"2015-12-02T11:17:54+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
{"reqId":"PlvOvONgwS52eFFC3gjn","remoteAddr":"::1","app":"PHP","message":"Illegal offset type in isset or empty at \/Users\/morrisjobke\/Projects\/owncloud\/master\/lib\/private\/user\/database.php#259","level":0,"time":"2015-12-02T11:17:54+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
{"reqId":"PlvOvONgwS52eFFC3gjn","remoteAddr":"::1","app":"PHP","message":"Undefined offset: 0 at \/Users\/morrisjobke\/Projects\/owncloud\/master\/lib\/private\/share\/mailnotifications.php#102","level":0,"time":"2015-12-02T11:17:54+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
{"reqId":"PlvOvONgwS52eFFC3gjn","remoteAddr":"::1","app":"core","message":"Sent mail to \"Array\n(\n    [def@ghi.de] => abc\n)\n\" with subject \"mjob hat \u201e\u201c mit Ihnen geteilt\"","level":0,"time":"2015-12-02T11:17:54+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
```

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 I still get following:

I hate missing unit tests :angry:

**Comment by MorrisJobke** [comment]
> I hate missing unit tests

I have a simple solution for this: Write them ;)

**Comment by DeepDiver1975** [comment]
> I have a simple solution for this: Write them ;)

you don't say that! :astonished:

File: lib/private/share/mailnotifications.php [hunk 2]
```diff
@@ -59,34 +57,31 @@ class MailNotifications {
...
-	 * @param IConfig $config
+	 * @param IUser $user
 	 * @param IL10N $l10n
 	 * @param IMailer $mailer
 	 * @param ILogger $logger
 	 * @param Defaults $defaults
 	 */
-	public function __construct($uid,
-								IConfig $config,
+	public function __construct($user,
```
Comment (nickvergessen) [inline lib/private/share/mailnotifications.php]: No typehint?

**Comment by nickvergessen** [comment]
I assume this is for LDAP etc, so they can store the email whereever they want?
Is there a way to get emails for a set of users, otherwise I will have to query each user manually in the activity app background job. (not that bad, since it's async, but still bad)

**Comment by DeepDiver1975** [comment]
> I assume this is for LDAP etc, so they can store the email whereever they want?
> Is there a way to get emails for a set of users, otherwise I will have to query each user manually in the activity app background job. (not that bad, since it's async, but still bad)

This is more about having a unified API. Functionality is unchanged.

File: lib/private/share/mailnotifications.php (hunk 2 above)
Comment (DeepDiver1975) [inline lib/private/share/mailnotifications.php]: no :see_no_evil:

File: apps/encryption/lib/crypto/encryptall.php [hunk 3]
```diff
@@ -361,8 +361,9 @@ protected function sendPasswordsByMail() {
 		foreach ($this->userPasswords as $recipient => $password) {
 			$progress->advance();
 			if (!empty($password)) {
-				$recipientDisplayName = $this->userManager->get($recipient)->getDisplayName();
-				$to = $this->config->getUserValue($recipient, 'settings', 'email', '');
+				$r = $this->userManager->get($recipient);
```
Comment (blizzz) [inline apps/encryption/lib/crypto/encryptall.php]: $r: improper variable name. It makes more sense calling $recipient $uid and $r → $recipient

**Comment by blizzz** [comment]
LDAP writes the email address to preferences table just as well. Still it's preferable and better maintainable to have just one spot to read it. Seems to work and code looks good to me, aside a variable naming thing. :+1:

**Comment by DeepDiver1975** [comment]
@MorrisJobke mind reviwing/testing again

**Comment by MorrisJobke** [comment]
Works so far, but there is one case that still throws an error (which isn't the case on master):

* create a user userA with email address set
* create a user userB **without** email address set
* share file "welcome.txt" of userB to userA and click on the "send email" text:

```
{"reqId":"xLvgmhVMlzfsNBPT5GQh","remoteAddr":"::1","app":"PHP","message":"Undefined offset: 1 at \/Users\/morrisjobke\/Projects\/owncloud\/master\/lib\/private\/mail\/message.php#62","level":0,"time":"2015-12-02T15:35:53+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
{"reqId":"xLvgmhVMlzfsNBPT5GQh","remoteAddr":"::1","app":"sharing","message":"Can't send mail to inform the user about an internal share: Address in mailbox given [@] does not comply with RFC 2822, 3.6.2.","level":3,"time":"2015-12-02T15:35:53+00:00","method":"POST","url":"\/master\/index.php\/core\/ajax\/share.php"}
```

**Comment by MorrisJobke** [comment]
Now the user management looks broken:

bildschirmfoto 2015-12-02 um 17 48 15


But sending the email works and all other tests are also fine :+1: once the above issue is fixed :see_no_evil:

**Pull request merged**